File: 2024/01/go/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"slices"
	"strconv"
	"strings"
)

type lists [][]int

var data = lists(make([][]int, 2))

func parse(r io.Reader) {
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		num0, _ := strconv.Atoi(fields[0])
		num1, _ := strconv.Atoi(fields[1])

		data[0] = append(data[0], num0)
		data[1] = append(data[1], num1)
	}
}

func (l lists) similarityScore() (res int) {
	counts := map[int]int{}

	for _, n := range l[1] {
		_, ok := counts[n]
		if !ok {
			counts[n] = 0
		}

		counts[n] += 1
	}

	for _, n := range l[0] {
		if qty, ok := counts[n]; ok {
			res += qty*n
		}
	}

	return
}

func (l lists) solve() (res int) {
	slices.Sort(l[0])
	slices.Sort(l[1])

	for i := range len(l[0]) {
		diff := l[0][i] - l[1][i]
		if diff < 0 {
			diff = -diff
		}

		res += diff
	}

	return
}

func main() {
	parse(os.Stdin)

	P1 := data.solve()
	P2 := data.similarityScore()

	fmt.Printf("P1: %d\n", P1)
	fmt.Printf("P2: %d\n", P2)
}

File: 2024/01/typescript/main.ts
import { readFileSync } from "node:fs"

let data: [Array<number>, Array<number>] = [[], []]
for (const line of readFileSync(0, { "encoding": "utf-8" }).split("\n")) {
    const [l0, l1] = line.split("   ")
    data[0].push(parseInt(l0))
    data[1].push(parseInt(l1))
}


// for some reason readFileSync adds an empty line at end, this removes it.
data[0].pop()
data[1].pop()

function solve(): number {
    data[0].sort((a, b) => a - b)
    data[1].sort((a, b) => a - b)

    let res = 0;
    for (let i = 0; i < data[0].length; i++) {
        res += Math.abs(data[0][i] - data[1][i])
    }

    return res
}

function similarityScore(): number {
    const counts = new Map<number, number>

    for (const i of data[1]) {
        const val = counts.get(i)
        if (!val) {
            counts.set(i, 1)
        } else {
            counts.set(i, val + 1)
        }
    }

    let res = 0
    for (const i of data[0]) {
        const qty = counts.get(i)
        if (qty) {
            res += qty * i
        }
    }

    return res
}

console.log(`P1: ${solve()}`)
console.log(`P2: ${similarityScore()}`)